src/uct/base/uct_iface.c

```diff
@@ -121,6 +123,22 @@ void uct_iface_dump_am(uct_base_iface_t *iface, uct_am_trace_type_t type,
     }
 }
 
+void uct_iface_mpool_empty_warn(uct_base_iface_t *iface, ucs_mpool_t *mp)
+{
+    static ucs_time_t warn_time = 0;
+    ucs_time_t now = ucs_get_time();
+
+    /* Limit the rate of warning to once in 30 seconds. This gives reasonable
+     * indication about a deadlock without flooding with warnings messages. */
+    if (warn_time == 0) {
+        warn_time = now;
+    }
+    if (now - warn_time > ucs_time_from_sec(30)) {
+        ucs_warn("Memory pool %s is empty", ucs_mpool_name(mp));
+        warn_time = now;
+    }
+}
+
```

Comment:
IC. when it's a constant state, it will be printed every 30sec and the job will not make progress. If it happens from time to time it will probably be printed only a few times and the job will complete.